js/marshal.go

```diff
@@ -0,0 +1,224 @@
+package js
+
+import (
+	"syscall/js"
+)
+
+type (
+	Type  int
+	Value struct {
+		_type        Type
+		_constructor string
+
+		value interface{}
+	}
+)
+
+const (
+	TypeNumber Type = iota
+	TypeBoolean
+	TypeString
+	TypeObject
+	TypeArray
+	TypeNull
+)
+
+func NewValue(value interface{}) *Value {
+	switch value.(type) {
+	case int, int32, int64, uint, uint32, uint64, float32, float64:
+		return &Value{
+			_type:        TypeNumber,
+			_constructor: "Number",
+			value:        value,
+		}
+	case bool:
+		return &Value{
+			_type:        TypeBoolean,
+			_constructor: "Boolean",
+			value:        value,
+		}
+	case string:
+		return &Value{
+			_type:        TypeString,
+			_constructor: "String",
+			value:        value,
+		}
+	case map[string]interface{}:
+		return &Value{
+			_type:        TypeObject,
+			_constructor: "Object",
+			value:        value,
+		}
+	case []interface{}:
+		return &Value{
+			_type:        TypeArray,
+			_constructor: "Array",
+			value:        value,
+		}
+	default:
+		return &Value{
+			_type:        TypeNull,
+			_constructor: "Null",
+			value:        nil,
+		}
+	}
+}
+
+func (v Value) Value() interface{} {
+	return v.value
+}
+
+func (v Value) Get(key string) interface{} {
+	val, ok := v.value.(map[string]interface{})[key]
+	if !ok {
+		return nil
+	}
+	return val.(Value).value
+}
+
+func (v Value) Set(key string, value interface{}) {
+	switch value.(type) {
+	case int, int32, int64, uint, uint32, uint64, float32, float64:
+		v.value.(map[string]interface{})[key] = Value{
+			_type:        TypeNumber,
+			_constructor: "Number",
+			value:        value,
+		}
+	case bool:
+		v.value.(map[string]interface{})[key] = Value{
+			_type:        TypeBoolean,
+			_constructor: "Boolean",
+			value:        value,
+		}
+	case string:
+		v.value.(map[string]interface{})[key] = Value{
+			_type:        TypeString,
+			_constructor: "String",
+			value:        value,
+		}
+	case map[string]interface{}:
+		v.value.(map[string]interface{})[key] = Value{
+			_type:        TypeObject,
+			_constructor: "Object",
+			value:        value,
+		}
+	case []interface{}:
+		v.value.(map[string]interface{})[key] = Value{
+			_type:        TypeArray,
+			_constructor: "Array",
+			value:        value,
+		}
+	default:
+		v.value.(map[string]interface{})[key] = Value{
+			_type:        TypeNull,
+			_constructor: "Null",
+			value:        nil,
+		}
+	}
+}
+
+func (v Value) String() string {
+	if v._type == TypeString {
+		return v.value.(string)
+	}
+	return ""
+}
+
+func (v Value) Type() Type {
+	return v._type
+}
+
+func (v Value) Constructor() string {
+	return v._constructor
+}
+
+// Unmarshal unmarshals a `syscall/js.Value` into the internal `Value` type.
+//
+// Note that only primitive types are supported and will be unmarshaled
+//   - Number
+//   - Boolean
+//   - String
+//   - Object (recursively converted to map[string]interface{})
+//   - Array (recursively converted to []interface{})
+func Unmarshal(v js.Value) *Value {
```

Comment:
This will need to be an interface... otherwise, we're going to get the same build error during testing :(

Reply:
Yes, I didn't realize that initially. However with an interface, we'll still come back to it, as we'll need to assert